### ISSUE TITLE
AWS SSO User with email Name.Surname@company.com not found

### ISSUE THREAD

**Issue by oleksandrsv** [fivexl/terraform-aws-sso-elevator#102]
Hi, sometimes we have an issue when lambda can not find user but he is present and has email in Slack and SSO.
Looks like lambda is looking as capital letters but in SSO they are small:

```
"level": "ERROR",
    "location": "exception:511",
    "message": "AWS SSO User with email Name.Surname@company.com not found",
    "timestamp": "2024-10-03 14:37:28,954+0000",
    "service": "errors",
    "exception": "Traceback (most recent call last):\n  File \"/var/task/errors.py\", line 42, in wrapper\n    return fn(*args, **kwargs)\n  File \"/var/task/main.py\", line 198, in handle_button_click\n    access_control.execute_decision(\n  File \"/var/task/access_control.py\", line 171, in execute_decision\n    user_principal_id = sso.get_user_principal_id_by_email(identitystore_client, sso_instance.identity_store_id, requester.email)\n  File \"/var/task/sso.py\", line 325, in get_user_principal_id_by_email\n    raise errors.NotFound(f\"AWS SSO User with email {email} not found\")\nerrors.NotFound: AWS SSO User with email Name.Surname@company.com not found",
    "exception_name": "NotFound",
```

AWS SSO:

```
Username | name.surname@company.com
-- | --
Email | name.surname@company.com Verified
```

Slack:

```
Name Surname
Active
5:56 PM local time
Contact information
Email Address
Name.Surname@company.com
```

**Comment by oleksandrsv** [comment]
```
def get_user_principal_id_by_email(client: IdentityStoreClient, identity_store_id: str, email: str) -> str:
    response = list_users(client, identity_store_id=identity_store_id)
    for user in response["Users"]:
        for user_email in user.get("Emails", []):
            if user_email.get("Value") == email:
                return user["UserId"]

    raise errors.NotFound(f"AWS SSO User with email {email} not found")
```

Maybe here we can add lower function?
```
name = "BOB STONE"
print(name.lower()) # >> bob stone
name1 = "Ruby Roundhouse"
print(name1.lower()) # >> ruby roundhouse
name2 = "joHN Wick"
print(name2.lower()) # >> john wick
name3 = "charlieNew"
print(name3.lower()) # >> charlienew
```

**Comment by EreminAnton** [comment]
Hello! Thanks for your feedback!

I do see an issue there. In the Lambda function, we simply get the requester's email from Slack (in this case `Name.Surname@company.com`) and try to find a user with the same email using the following function:

```python
def get_user_principal_id_by_email(client: IdentityStoreClient, identity_store_id: str, email: str) -> str:
    response = list_users(client, identity_store_id=identity_store_id)
    for user in response["Users"]:
        for user_email in user.get("Emails", []):
            if user_email.get("Value") == email:
                return user["UserId"]

    raise errors.NotFound(f"AWS SSO User with email {email} not found")
```
The comparison is done using this line: `if user_email.get("Value") == email`:, which uses `==` for the comparison.

When we test this comparison, we get False as a result:

```python
a = "Name.Surname@company.com"
b = "name.surname@company.com"
print(a == b)  # This returns False
```
From a code perspective, the fix for this issue would be easy to implement by using:

```python
print(a.lower() == b.lower())  # This returns True
```
Since AWS SSO is not case-sensitive (i.e., we can't create users with emails that differ only by uppercase or lowercase), I think it would be fine to just add the `.lower()` function. I'll discuss this with the team to see if there are any other security concerns I might have missed and get back to you.

As a temporary solution, you can rename the SSO user to match the case of the Slack user, which should fix the problem.

**Comment by EreminAnton** [comment]
@oleksandrsv

I've implemented a fix in Release 2.0.2 that should resolve the case sensitivity issue with email comparisons during AWS SSO user lookups.

The email comparison is now case-insensitive, aligning with AWS SSO's behavior.

If you encounter any further issues, feel free to reach out.
